### PR TITLE
Support OpenOceanV2 on Metis

### PR DESCRIPTION
**Pull request by Jeff-CCH**
- [x] OpenOpcean logic
- [x] test
- [x] yarn changeset

### CODE REVIEW

File: src/logics/openocean-v2/logic.swap-token.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  implements core.LogicTokenListInterface, core.LogicOracleInterface, core.LogicBuilderInterface
+{
+  static readonly supportedChainIds = supportedChainIds;
+
+  async getTokenList() {
+    const resp = await axios.get(`https://open-api.openocean.finance/v3/${this.chainId.toString()}/tokenList`);
+    const tokens = resp.data.data;
+    const tokenList: SwapTokenLogicTokenList = [];
+    for (const { chain, address, decimals, symbol, name } of tokens) {
+      tokenList.push(new common.Token(getChainId(chain), address, decimals, symbol, name));
```
Comment (chouandy) [inline src/logics/openocean-v2/logic.swap-token.ts:33]: Just use this.chainId

File: src/logics/openocean-v2/logic.swap-token.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+@core.LogicDefinitionDecorator()
+export class SwapTokenLogic
+  extends core.Logic
+  implements core.LogicTokenListInterface, core.LogicOracleInterface, core.LogicBuilderInterface
+{
+  static readonly supportedChainIds = supportedChainIds;
+
+  async getTokenList() {
+    const resp = await axios.get(`https://open-api.openocean.finance/v3/${this.chainId.toString()}/tokenList`);
```
[the remark this answers is not in the document]
Comment (chouandy) [reply src/logics/openocean-v2/logic.swap-token.ts:29]: no need `toString()`

File: src/logics/openocean-v2/configs.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    accumulator[1][config.chainId] = config;
+    return accumulator;
+  },
+  [[], {}] as [number[], Record<number, Config>]
+);
+
+export function getChainId(chain: string) {
+  if (chain === 'metis') return common.ChainId.metis;
+  else return 0;
+}
```
Comment (chouandy) [inline src/logics/openocean-v2/configs.ts:29]: no need this function

File: src/logics/openocean-v2/logic.swap-token.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  // If you wish to exclude quotes from a specific DEX, you can include the corresponding DEX ID
+  // in the 'disabledDexIds' parameter. You can retrieve the DEX IDs from the following API:
+  // https://open-api.openocean.finance/v3/{chainId}/dexList
+  async quote(params: SwapTokenLogicParams) {
+    const { input, tokenOut, disabledDexIds } = params;
+    const gasPrice = getGasPrice(this.chainId);
+    let slippage = slippageToOpenOcean(params.slippage ?? 0);
+
+    const client = axios.create({ baseURL: `https://open-api.openocean.finance/v3/${this.chainId.toString()}` });
+    const resp = await client.get(`/quote`, {
```
Comment (chouandy) [inline src/logics/openocean-v2/logic.swap-token.ts:47]: just

```suggestion
    const resp = await axios.get(`https://open-api.openocean.finance/v3/${this.chainId}/quote`, {
```

File: src/logics/openocean-v2/logic.swap-token.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      },
+    });
+
+    const { to, data } = resp.data.data;
+    const approveTo = getExchangeAddress(this.chainId);
+    const tokenInput = new common.TokenAmount(
+      new common.Token(this.chainId, input.token.address, input.token.decimals, input.token.symbol, input.token.name),
+      input.amount
+    );
+    const inputs = [core.newLogicInput({ input: tokenInput })];
```
Comment (chouandy) [inline src/logics/openocean-v2/logic.swap-token.ts:94]:
```suggestion
    const inputs = [core.newLogicInput({ input })];
```

File: src/logics/openocean-v2/logic.swap-token.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  // The best overall return =  out_value - tx cost and the tx_cost = gas_used & gas_price
+  async build(fields: SwapTokenLogicFields, options: SwapTokenLogicOptions) {
+    const { input, output, disabledDexIds } = fields;
+    const { account } = options;
+    const gasPrice = getGasPrice(this.chainId);
+    const agent = await this.calcAgent(account);
+    const slippage = slippageToOpenOcean(fields.slippage ?? 0);
+
+    const client = axios.create({ baseURL: `https://open-api.openocean.finance/v3/${this.chainId.toString()}` });
+    const resp = await client.get(`/swap_quote`, {
```
Comment (chouandy) [inline src/logics/openocean-v2/logic.swap-token.ts:76]:
```suggestion
    const resp = await axios.get(`https://open-api.openocean.finance/v3/${this.chainId}/swap_quote`, {
```

File: src/logics/openocean-v2/slippage.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+// TODO: test floating point
+export function slippageToOpenOcean(slippage: number) {
+  const min = 0.05;
+  const max = 50;
+  const normalizedSlippage = slippage / 100;
+  if (normalizedSlippage < min) return min;
+  if (normalizedSlippage > max) return max;
+  return normalizedSlippage;
+}
```
Comment (chouandy) [inline src/logics/openocean-v2/slippage.ts:9]:
```suggestion
export function slippageToOpenOcean(slippage: number) {
  return Math.min(Math.max(slippage / 100, 0.05), 50);
}
```

File: src/logics/openocean-v2/logic.swap-token.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
+import { axios } from 'src/utils';
+import * as common from '@protocolink/common';
+import * as core from '@protocolink/core';
+import { getApiUrl } from '.';
+import { getGasPrice, supportedChainIds } from './configs';
```
Comment (zodahu) [inline src/logics/openocean-v2/logic.swap-token.ts:5]:
```suggestion
import { getApiUrl, getGasPrice, supportedChainIds } from './configs';
```

Comment (Jeff-CCH) [reply src/logics/openocean-v2/logic.swap-token.ts]: fixed.

**Review by zodahu — APPROVED** [review]
LGTM

**Review by chouandy — APPROVED** [review]
LGTM

**Pull request merged**